nbdime/prettyprint.py

```diff
@@ -1,355 +1,540 @@
+import datetime
 import pprint
 import re
 import shutil
 from subprocess import Popen, PIPE
 import tempfile
 from difflib import unified_diff
-
-try:
-    from textwrap import indent
-except ImportError:
-    def indent(text, prefix):
-        """The relevant part of textwrap.indent for Python 2"""
-        return prefix + text.replace('\n', '\n' + prefix)
-
 from six import string_types
+import hashlib
 
-import colorama
-
-from .diff_format import NBDiffFormatError, DiffOp
-from .patching import patch
-from .utils import star_path, split_path, join_path
+#try:
+#    from textwrap import indent
+#except ImportError:
+#    def indent(text, prefix):
+#        """The relevant part of textwrap.indent for Python 2"""
+#        return prefix + text.replace('\n', '\n' + prefix)
 
 try:
     from shutil import which
 except ImportError:
     from backports.shutil_which import which
 
-# Toggle indentation here
-with_indent = False
+from .diff_format import NBDiffFormatError, DiffOp, op_patch
+from .patching import patch, patch_string
+from .utils import star_path, split_path, join_path
 
-# Change to enable/disable color print etc.
-_git_diff_print_cmd = 'git diff --no-index --color-words'
+# TODO: Make this configurable
+use_git = True
+use_diff = True
+use_colors = True
+
+# Indentation offset in pretty-print
+IND = "  "
+
+# Max line width used some placed in pretty-print
+# TODO: Use this for line wrapping some places?
+MAXWIDTH = 78
+
+if use_colors:
+    import colorama
+    RED = colorama.Fore.RED
+    GREEN = colorama.Fore.GREEN
+    BLUE = colorama.Fore.BLUE
+    RESET = colorama.Style.RESET_ALL
+    _git_diff_print_cmd = 'git diff --no-index --color-words'
+else:
+    RED = ''
+    GREEN = ''
+    BLUE = ''
+    RESET = ''
+    _git_diff_print_cmd = 'git diff --no-index'
+
+REMOVE   = '{color}-  '.format(color=RED)
+ADD      = '{color}+  '.format(color=GREEN)
+INFO     = '{color}## '.format(color=BLUE)
+DIFF_ENTRY_END = '\n'
+
+def _external_diff_render(cmd, a, b):
+    try:
+        if isinstance(a, bytes):
+            a = a.decode("utf8")
+        if isinstance(b, bytes):
+            b = b.decode("utf8")
+        td = tempfile.mkdtemp()
+        with io.open(os.path.join(td, 'before'), 'w', encoding="utf8") as f:
+            f.write(a)
+        with io.open(os.path.join(td, 'after'), 'w', encoding="utf8") as f:
+            f.write(b)
+        p = Popen(cmd + ['before', 'after'], cwd=td, stdout=PIPE)
+        out, _ = p.communicate()
+        diff = out.decode('utf8')
+    finally:
+        shutil.rmtree(td)
+    return diff
+
+
+def _builtin_diff_render(a, b):
+    gen = unified_diff(
+        a.splitlines(False),
+        b.splitlines(False),
+        lineterm='')
+    uni = []
+    for line in gen:
+        if line.startswith('+'):
+            uni.append("%s%s%s" % (ADD, line[1:], RESET))
+        elif line.startswith('-'):
+            uni.append("%s%s%s" % (REMOVE, line[1:], RESET))
+        else:
+            uni.append(line)
```

Comment:
Done. Also dropping the "No newline at end of file", it doesn't make sense to me here.